[doc] Improve type system documentation (#4002)
taichi-dev/taichi: docs/lang/articles/basic/type.md

- `u` for unsigned integers, e.g. 128, 256
- `f` for floating point numbers, e.g. 3.14, 1.0, 1e-4

The _digital number_ can be either `8`, `16`, `32`, or `64`.
It represents how many **bits** are used in storing the data. The larger

> the bit number is, the higher the precision is.

Author: Why, can you provide a reference please? AFAIK, the two clauses should match. Either all with `is`, or without. With `is` each is a sentence, so i think the changed is better?

> the bit number is, the higher the precision is.

Reviewer: Oops. Missed this one. You are right. Please go ahead with your version.

### Supported primitive types

Currently in Taichi, the supported primitive types on each backend are:

| type | CPU/CUDA |  OpenGL | Metal |  Vulkan  |
| ---- | -------- | ------- | ----- | -------- |
| i8   |:heavy_check_mark:|:x:                   |:heavy_check_mark:|:large_orange_diamond:|
| i16  |:heavy_check_mark:|:x:                   |:heavy_check_mark:|:large_orange_diamond:|
| i32  |:heavy_check_mark:|:heavy_check_mark:    |:heavy_check_mark:|:heavy_check_mark:    |
| i64  |:heavy_check_mark:|:large_orange_diamond:|:x:               |:large_orange_diamond:|
| u8   |:heavy_check_mark:|:x:                   |:heavy_check_mark:|:large_orange_diamond:|
| u16  |:heavy_check_mark:|:x:                   |:heavy_check_mark:|:large_orange_diamond:|
| u32  |:heavy_check_mark:|:x:                   |:heavy_check_mark:|:heavy_check_mark:    |
| u64  |:heavy_check_mark:|:x:                   |:x:               |:large_orange_diamond:|

> | f16  |:heavy_check_mark:|:x:                   |:x:               |:heavy_check_mark:    |

Reviewer 2: ```suggestion
```
It looks like we don't have this type

| f32  |:heavy_check_mark:|:heavy_check_mark:    |:heavy_check_mark:|:heavy_check_mark:    |
| f64  |:heavy_check_mark:|:heavy_check_mark:    |:x:               |:large_orange_diamond:|

> (:large_orange_diamond: Requiring extension.)

Author: sure, any suggestions?

> (:large_orange_diamond: requires extension)

Reviewer 2: ```suggestion
(:large_orange_diamond: means extension required)
```

> (:large_orange_diamond: means extension required)

Author: @mzmzm could you please fix this?

> (:large_orange_diamond: means extension required)

Reviewer 2: commit ba1bf56de845638137a8129f6002ffa83621d974

    ...
```

### Type promotion

When Taichi performs binary operations on different types, the result is a

> promoted type. This is known as type promotion. Following the C programming

Reviewer: This didn't explain what a promoted type is. Please could you provide more details?

language convention, Taichi always chooses the more precise type to contain the
result value. For example:

- `i32 + f32 = f32` (integer + float = float)
- `i32 + i64 = i64` (less-bits + more-bits = more-bits)

### Type casts

When using Taichi, you may encounter situations when certain type of data needs
to be assigned to another type. In this case, type casts are needed. Here, we
list four use cases commonly seen during Taichi programming, namely implicit
casts, explicit casts, casting vectors and matrices, and bit-casts.

:::caution
In Taichi, the type of a variable is **determined on its initialization**.

> All data types are static in the **Taichi scope**.

Reviewer 2: ```suggestion
All data types are static in the **Taichi scope**. (A static type of variable means the type is known at compile time.  This lets the compiler decide whether a given variable can perform the actions requested from it or not. For more information, see [the difference between statically typed and dynamically typed languages](https://stackoverflow.com/questions/1517582/what-is-the-difference-between-statically-typed-and-dynamically-typed-languages)
```

Author: @mzmzm thanks, can you adapt the sentence a bit?
1: (Static here means...... Based on this information, the compiler can decide whether the performed action is legal or not.). 
2: please don't use stackoverflow as an official reference in our document. Wikipedia is fine. cppreference.com is fine. but not random post. Can you try to find another good reference to be linked?

Reviewer 2: Thanks. I update my suggestion.

Reviewer 3: Did you forget to apply suggestions here @qiao-bo ?

:::

> #### Implicit casts

Reviewer: Implicit cast? 
Singular form or plural form?

Author: Not sure, i am fine as long as they are consistent per level title :/

When a _low-precision_ variable is assigned to a _high-precision_

> variable, it will be implicitly promoted to the _high-precision_ type

Reviewer: Present tense: 

, it is implicitly promoted to the _high-precision_ type without any warning.

    ...
```

### Type promotion

When Taichi performs binary operations on different types, the result is a
promoted type. This is known as [type promotion](https://en.wikipedia.org/wiki/Type_conversion#Type_promotion).

> Following the C programming

Reviewer 4: nit: Can we add a link to the C spec?

Author: @strongoier ;/

    a = 1
    print(a)  # 1.0
```
In this example, variable `a` is initialized with `float` type. On the next line, the assign statement will cast `1` from `int` type to `float` type implicitly.

> 

Reviewer: In this example, variable `a` is initialized with a `float` type. On the next line, the assignment statement casts `1` from `int` to `float` implicitly.

    a = 3.14
    print(a)  # 3
```

> In this example, variable `a` is initialized with `int` type. On the next line, the assign statement will cast `3.14` from `float` type to `int` type implicitly.

Reviewer: ```suggestion
In this example, variable `a` is initialized with an `int` type. On the next line, the assignment statement casts `3.14` from `float` to `int` implicitly.
```

my_mat2f = ti.types.matrix(2, 2, float)
my_ray3f = ti.types.struct(ro=my_vec3f, rd=my_vec3f, l=ti.f32)
```

> In this example, we define four compound types which can be used to create fields and create local variables. We can learn how to use it below.

Reviewer: ```suggestion
In this example, we define four compound types for creating fields and local variables. 
```

mat2 = ti.Matrix.field(2, 2, dtype=ti.i32, shape=(24, 32))
ray3 = ti.Struct.field({'ro': my_vec3f, 'rd': my_vec3f, 'l': ti.f32}, shape=(1024, 768))
```

> In this example, we define three fields which have different types and shapes by two ways. They are exactly the same.

Reviewer: ```suggestion
In this example, we define three fields in two different ways but of exactly the same effect.
```
